### ISSUE TITLE
Project references ignore module type

### ISSUE THREAD

**Issue by TimvdLippe** [microsoft/TypeScript#32554]
**TypeScript Version:**  3.5.3

**Search Terms:** project references commonjs

**Code**

This assumes `@types/node` is installed

tsconfig.json
```json
{
  "compilerOptions": {
    "target": "esnext",
    "composite": true,
    "strict": true,
    "esModuleInterop": true,
    "moduleResolution": "node"
  },
  "references": [{
    "path": "./node/"
  }]
}
```

node/tsconfig.json
```json
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "module": "commonjs"
  }
}
```

node/bug.ts
```ts
import path from 'path';

console.log(path);
```

Run `tsc -b` in the root folder. This generates a `bug.js` with the expected output. Now add a space to `bug.ts` (just force a change in the file, doesn't matter what). Run `tsc -b` again.

**Expected behavior:**
The output of node/bug.js should be
```js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
console.log(path_1.default);
```

**Actual behavior:**
```js
import path from 'path';
console.log(path);
```

It seems that the module type is ignored once the `tsconfig.tsbuildinfo` is generated. Deleting `tsconfig.tsbuildinfo` and running `tsc -b` results in the expected behavior.

**Playground Link:** n/a

**Related Issues:** I could not find any open/closed issues that have a similar problem

**Comment by sheetalkamat** [comment]
This is working as intended because your project `tsconfig.json` is not empty project. It compiles same files as `node/tsconfig.json` resulting in the build of same project again. Here are files emitted

```
C:\temp\test>node c:\TypeScript\built\local\tsc.js -b -v --listEmittedFiles
[1:11:46 PM] Projects in this build:
    * node/tsconfig.json
    * tsconfig.json

[1:11:46 PM] Project 'node/tsconfig.json' is out of date because output file 'node/bug.js' does not exist

[1:11:46 PM] Building project 'C:/temp/test/node/tsconfig.json'...

TSFILE: C:/temp/test/node/bug.js
TSFILE: C:/temp/test/node/bug.d.ts
TSFILE: C:/temp/test/node/tsconfig.tsbuildinfo
[1:11:48 PM] Project 'tsconfig.json' is out of date because output file 'tsconfig.tsbuildinfo' does not exist

[1:11:48 PM] Building project 'C:/temp/test/tsconfig.json'...

TSFILE: C:/temp/test/node/bug.js
TSFILE: C:/temp/test/node/bug.d.ts
TSFILE: C:/temp/test/tsconfig.tsbuildinfo
```

**Comment by TimvdLippe** [comment]
Thanks for the detailed explanation. That said, we think this behavior is very confusing. Would you consider changing it such that a project with project references automatically excludes these files?

E.g. a reference with `"path": "./node/"` implicitly excludes `./node/` from building (since that is a reference and the root config should not recompile projects twice).

When I try to make the exclusions explicit, our sub-projects (that extend the root tsconfig) will now ignore their content because we explicitly exclude these path references.

TLDR: A root tsconfig should implicitly exclude sources in path references, as they are inherently part of a different project. They can't be explicitly excluded, because the tsconfig extend each other.

**Comment by TimvdLippe** [comment]
Friendly ping @sheetalkamat We are constantly running into the issue where root projects are recompiling all of our nested projects, unless we explicitly add `"exclude"` as well as duplicate all of our other `tsconfig.json` (like `compilerOptions`).

**Comment by sheetalkamat** [comment]
Files included by the program are irrespective of project references and that calculation takes place way before we even dive deeper into what project references are part of program. I don't think we want to depend on project references and what those files include to determine whether to include them. You need to explicitly include what you don't want in your program as part of root files

**Comment by TimvdLippe** [comment]
> You need to explicitly include what you don't want in your program as part of root files

I understand that this is required, but as stated above, this does not work if you are extending configurations. E.g. if you want to have a unified configuration for strict checks, but want to differentiate on `compilerOptions.module`, you will run into this issue.

I also acknowledge that the way the project references are implemented prohibits from moving the exclusion calculation, but I would hope that that can be addressed in a future architectural update.

Right now, the behavior for the user is confusing, less performant (because it is really hard to figure out that TS is actually compiling files multiple times) and require extensive duplication of TS configuration options like `compilerOptions`.
